test/extended/apiserver/webhook.go

```diff
@@ -0,0 +1,386 @@
+package apiserver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+
+	exetcd "github.com/openshift/origin/test/extended/etcd"
+	exutil "github.com/openshift/origin/test/extended/util"
+
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
+	discocache "k8s.io/client-go/discovery/cached"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+	e2edeployment "k8s.io/kubernetes/test/e2e/framework/deployment"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	"k8s.io/utils/pointer"
+)
+
+var _ = g.Describe("[sig-api-machinery][Feature:AdmissionWebhook]", func() {
+	defer g.GinkgoRecover()
+	oc := exutil.NewCLI("apiserver")
+
+	g.It("Validating Admission Webhook should be called when accessing OpenShift APIs", func() {
+		targetNamespace := oc.Namespace()
+		kubeClient := oc.AdminKubeClient()
+		servicePort := int32(8443)
+		containerPort := int32(8444)
+		mapper := restmapper.NewDeferredDiscoveryRESTMapper(discocache.NewMemCacheClient(kubeClient.Discovery()))
+		adminRestConfig := oc.AdminConfig()
+
+		annotateNamespace(kubeClient, targetNamespace, "webhook-marker")
+		createService(kubeClient, targetNamespace, "openshift-test-webhook", "openshift-test-webhook", servicePort, containerPort)
+		deployWebhookServer(kubeClient, imageutils.GetE2EImage(imageutils.Agnhost), targetNamespace, "openshift-test-webhook", "openshift-test-webhook", "openshift-test-webhook", containerPort)
+		removeWebhookConfiguration := createValidatingWebhook(kubeClient, targetNamespace, "webhook-marker", "webhook-marker", "openshift-test-webhook", "openshift-test-webhook", servicePort)
+		defer removeWebhookConfiguration()
+
+		storageData := exetcd.OpenshiftEtcdStorageData
+		for key := range storageData {
+			gvr := key
+			data := storageData[gvr]
+
+			// apply for core types is already well-tested, so skip
+			// openshift types that are just aliases.
+			aliasToCoreType := data.ExpectedGVK != nil
```

Comment:
we could reuse the code with the test that @marun prepared but it seems to be more readable as it is now